### PR TITLE
Add `benchmark-block` command

### PR DESCRIPTION
**Pull request by ggwpez**
Deja vu! The code that I originally wanted to use for #10977 is now put to better use.  

This command helps in gauging the precision or our benchmarking.  
It compares the time that it takes to execute a block to its consumed weight. Ideally they are the same.  

Example with a CPU intensive hand-crafted block where our benchmarking worked well:  
```rust
Block 2 with     2 tx used 102.39% of its weight ( 1,281,193,264 of 1,251,309,762  ns) - OVER WEIGHT!
```

Or some empty blocks where we over-estimate:
```rust
Block 1 with     1 tx used  65.03% of its weight (     4,464,209 of      6,864,645 ns)    
Block 2 with     1 tx used  65.15% of its weight (     4,472,344 of      6,864,645 ns)    
Block 3 with     1 tx used  64.27% of its weight (     4,492,493 of      6,989,645 ns)
```

There are rust docs on the `BlockCmd` type with an example on how to invoke it.  
In the future this could be used to setup some live monitoring where each day we re-execute the blocks of the past 24h to see which ones are over weight (bad).  

Follow ups:  
- [ ] Polkadot integration
- [ ] Add a minimal test
- [ ] Save all results as JSON for further analysis, with split up into mandatory/operational weight
- [ ] Add `--block-weight-key` with default `26aa394eea5630e07c48ae0c9558cef734abf5cb34d6244378cddbf18e849d96`
- [ ] optional: Some convenience flags such as `from` and `to` can be negative to go into the past from the best block.

### CODE REVIEW

File: utils/frame/benchmarking-cli/src/block/cmd.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+use std::{fmt::Debug, sync::Arc};
+
+use super::bench::{Benchmark, BenchmarkParams};
+
+/// Benchmark the execution time historic blocks.
+///
+/// This can be used to verify that blocks do not use more weight than they consumed
+/// in their `WeightInfo`. Example:
+///
+/// Let's say you are on a Substrate chain and want to verify that the last 3 blocks
```
Comment (shawntabrizi) [inline utils/frame/benchmarking-cli/src/block/cmd.rs:37]:
```suggestion
/// Let's say you are on a Substrate chain and want to verify that the first 3 blocks
```

Comment (shawntabrizi) [inline utils/frame/benchmarking-cli/src/block/cmd.rs:37]:
```suggestion
/// Let's say you are on a Substrate chain and want to verify that the first 3 blocks
```

File: utils/frame/benchmarking-cli/src/block/cmd.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+/// did not use more weight than declared which would otherwise be an issue.
+/// To test this with a dev node, first create one with a temp directory:
+///
+/// $ substrate --dev -d /tmp/my-dev --execution wasm --wasm-execution compiled
+///
+/// And wait some time to let it 3 blocks. Then benchmark them with:
```
Comment (shawntabrizi) [inline utils/frame/benchmarking-cli/src/block/cmd.rs:43]:
```suggestion
/// And wait some time to let it produce 3 blocks. Then benchmark them with:
```

Comment (shawntabrizi) [inline utils/frame/benchmarking-cli/src/block/cmd.rs:43]:
```suggestion
/// And wait some time to let it produce 3 blocks. Then benchmark them with:
```

File: utils/frame/benchmarking-cli/src/block/bench.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+		Ok(())
+	}
+
+	/// Return the average *execution* aka. *import* time of the block.
+	fn measure_block(&self, block: &Block, parent_num: &BlockId<Block>) -> Result<NanoSeconds> {
+		let mut record = Vec::<NanoSeconds>::default();
+		// Interesting part here:
+		// Execute the block multiple times and collect stats about its execution time.
+		for _ in 0..self.params.repeat {
```
Comment (shawntabrizi) [inline utils/frame/benchmarking-cli/src/block/bench.rs:111]: did you check if the repeat times are all pretty much the same. I might be worried like the first time is slow, but then the subsequent runs are fast

Comment (shawntabrizi) [inline utils/frame/benchmarking-cli/src/block/bench.rs:111]: did you check if the repeat times are all pretty much the same. I might be worried like the first time is slow, but then the subsequent runs are fast

Comment (shawntabrizi) [reply utils/frame/benchmarking-cli/src/block/bench.rs:111]: if there is some weird caching behavior happening

Comment (shawntabrizi) [reply utils/frame/benchmarking-cli/src/block/bench.rs:111]: if there is some weird caching behavior happening

**Review by shawntabrizi — APPROVED** [review]
this is way better than i hoped!

File: utils/frame/benchmarking-cli/src/block/bench.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		Ok(took)
+	}
+
+	/// Returns the total nanoseconds of a [`frame_system::ConsumedWeight`] for a block number.
+	///
+	/// This is the post-dispatch corrected weight and is only available
+	// after executing the block.
+	fn consumed_weight(&self, block: &BlockId<Block>) -> Result<NanoSeconds> {
+		// Hard-coded key for System::BlockWeight. It could also be passed in as argument
+		// for the benchmark, but I think this should work as well.
```
Comment (kianenigma) [inline utils/frame/benchmarking-cli/src/block/bench.rs:133]: Or you can make it a new runtime api function in the custom benchmarking api?

**Review by kianenigma — APPROVED** [review]
LGTM, but a minor existential question I have is that why is this not part of the same `check-block` command that we have that is capable of re-importing old blocks? if you run with `-lruntime=debug`, you already get the consumed weight that was recorded onchain, and you can time and compare it with the `check-block` command again. 

I can see some benefit in having a custom command be easier to deal with and extend, but was the alternative considered?

**Comment by ggwpez** [comment]
Very good question! I initially had a look at that code, but did not think about extending it.  

Analyzing again; `check-block` uses the import queue, and that has a fast path, just returning with `Block already in chain`.  
Visible with `-lsync=trace`:
```pre
substrate --dev -d /tmp/dev
substrate check-block 1 --dev -d /tmp/dev --pruning archive -lsync=trace

# prints
Scheduling 1 blocks for import    
Starting import of 1 blocks  (1)    
Header 0x7e5d…ca03 has 3 logs    
Block already in chain 1: 0x7e5d3a...
Block imported successfully Some(1) (0x7e5d…ca03) 
```

So it does not actually re-execute the block, but returns quickly since the block is already known-good.

File: utils/frame/benchmarking-cli/src/block/bench.rs (hunk 3 above)
Comment (shawntabrizi) [reply utils/frame/benchmarking-cli/src/block/bench.rs:111]: As mentioned in our 1:1, lets get a bunch of metrics like min, max, std deviation etc with an extra flag like `--raw`

File: utils/frame/benchmarking-cli/src/block/bench.rs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		took: NanoSeconds,
+	) {
+		// The ratio of weight that the block used vs the number that it consumed.
+		// This should in general not exceed 100% (minus outliers).
+		let percent = (took as f64 / consumed as f64) * 100.0;
+		let num = if let BlockId::Number(num) = num {
+			num
+		} else {
+			unreachable!("Only put numbers in; qed")
+		};
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/src/block/bench.rs:163]: Why then take `BlockId`?

File: utils/frame/benchmarking-cli/src/block/bench.rs (hunk 4 above)
Comment (bkchr) [reply utils/frame/benchmarking-cli/src/block/bench.rs:133]: Yeah, this should be at least configurable.

File: utils/frame/benchmarking-cli/src/block/bench.rs (hunk 5 above)
Comment (ggwpez) [reply utils/frame/benchmarking-cli/src/block/bench.rs:163]: Ah yea, had the code structured differently before. Will change.

File: utils/frame/benchmarking-cli/src/block/bench.rs (hunk 4 above)
Comment (ggwpez) [reply utils/frame/benchmarking-cli/src/block/bench.rs:133]: Okay I will put all of that in a follow up issue to add either:  
A `--blockweight-key` parameter or do a `BenchmarkingApi` for the runtime.

**Comment by ggwpez** [comment]
bot rebase

**Comment by ggwpez** [comment]
bot rebase

**Comment by ggwpez** [comment]
bot merge

**Pull request merged**